Optionally load weights by name

arel opened the pull request:
This PR adds the **option to load model weights by layer name** rather than by flattened layer-order. This is intended to help when fine-tuning or transfer-learning a pretrained model. The alternative as far as I can tell is to load the weights into the *original* architecture, then surgically alter the layers. For Functional API models (not Sequential models) this gets more tricky, and I didn't find examples for altering such models in the Keras documentation or elsewhere.

The proposal is to add a boolean argument, `by_name` (default `False`), to the `load_weights()` method of a model. If set to `True`, then any layers of the new model with names matching those of the old model will be loaded. Layers without names, or layers with differing names are ignored.

This lets you alter and re-purpose an existing model by simply renaming the layers you want to change.

Comments are welcome.

### To do

 - [x] add unit tests
 - [x] batch-load weights
 - [x] update documentation with examples
 - [x] rebase and squash commits

fchollet commented on keras/engine/topology.py:
```diff
@@ -2550,6 +2554,69 @@ def load_weights_from_hdf5_group(self, f):
+                if layer.name:
+                    index.setdefault(layer.name, []).append(layer)
+
+            # We do not batch weight value assignments when assigning by name.
```
You should, though.

fchollet commented on keras/engine/topology.py:
```diff
@@ -2550,6 +2554,69 @@ def load_weights_from_hdf5_group(self, f):
+                                        ' expects ' + str(len(symbolic_weights)) +
+                                        ' weights, but the saved weights have ' +
+                                        str(len(weight_values)) +
+                                        ' elements.')
```
By definition `name` and `layer.name` will be the same. The error message should be modified to reflect this.

fchollet commented on keras/engine/topology.py:
```diff
@@ -2550,6 +2554,69 @@ def load_weights_from_hdf5_group(self, f):
+            index = {}
+            for layer in flattened_layers:
+                if layer.name:
+                    index.setdefault(layer.name, []).append(layer)
```
You don't need to compute a list of filtered layers in this setup. You can build the index in one go.

fchollet commented:
That is a good feature to add.

arel commented on tests/test_model_saving.py:
```diff
@@ -161,5 +162,110 @@ def test_saving_right_after_compilation():
+
+    out = model.predict(x)
+    old_weights = [layer.get_weights() for layer in model.layers]
+    _, fname = tempfile.mkstemp('.h5')
```
I think the [`tempfile.mkstemp()`](https://docs.python.org/2/library/tempfile.html#tempfile.mkstemp) is a more robust way of generating temp files in a platform agnostic way (instead of `np.random`). Thoughts about replacing the other occurrences for consistency? (Or reverting this one?)

fchollet replied:
That's right, we could use tempfiles across all tests.

arel replied:
Sounds good, I'll make this change.

arel commented:
Thanks, @fchollet! I addressed your comments and added a couple unit tests. I'm planning on updating the documentation with an example of this feature, but other than that I think it's ready.

arel commented:
This is ready to merge on my end.

I updated the documentation, squashed the commits and tests pass. If there's anything else you'd like me to change let me know.

Thanks!

arel commented:
Hi, @fchollet. Just a quick ping. I believe this PR is ready to merge.

fchollet commented on keras/engine/topology.py:
```diff
@@ -2469,14 +2469,27 @@ def save_weights_to_hdf5_group(self, f):
         '''Load all layer weights from a HDF5 save file.
+
+        If `by_name` is False (default) weights are loaded
+        in flattened array order, meaning the architecture
```
"weights are loaded based on the network's topology" might be a clearer way to put it.

fchollet commented on keras/engine/topology.py:
```diff
@@ -2552,6 +2565,54 @@ def load_weights_from_hdf5_group(self, f):
 
+    def load_weights_from_hdf5_group_by_name(self, f):
+        '''Weight loading is based on layer names
+        (not by sequential order).
```
"Name-based weight loading (instead of topological weight loading)"

fchollet commented on keras/engine/topology.py:
```diff
@@ -2552,6 +2565,54 @@ def load_weights_from_hdf5_group(self, f):
+        if 'nb_layers' in f.attrs:
+                raise Exception('You are trying to load a weight file ' +
+                                'by name with an unsupported legacy' +
+                                'format.')
```
"The weight file you are trying to load is in a legacy format that does not support name-based weight loading".

fchollet commented on keras/engine/topology.py:
```diff
@@ -2552,6 +2565,54 @@ def load_weights_from_hdf5_group(self, f):
+                    if len(weight_values) != len(symbolic_weights):
+                        raise Exception('Layer #' + str(k) +
+                                        ' (named "' + layer.name +
+                                        '" in the current model) expects ' +
```
Singe the names are expected to match across models, "in the current model" does not make sense here.

arel commented:
I made the changes you asked for. Anything else?

arel commented:
@fchollet, fixed merge conflict; ready for review/merge.

fchollet commented on docs/templates/getting-started/faq.md:
````diff
@@ -113,12 +113,39 @@ Note that you will first need to install HDF5 and the Python library h5py, which
 model.load_weights('my_model_weights.h5')
 ```
 
+If you need to load weights into a *different* archiecture (with some layers in common), for instance if fine-tuning or transfer-learning, you can load weights by *layer name*:
````
architecture (typo). "for" fine-tuning.

fchollet commented on docs/templates/getting-started/faq.md:
```diff
@@ -113,12 +113,39 @@ Note that you will first need to install HDF5 and the Python library h5py, which
+# new model
+model = Sequential()
+model.add(Dense(2, input_dim=3, name="dense_1"))  # will be loaded
+model.add(Dense(10, name="my_new_embedding"))  # will not be loaded
```
Please use reasonable names, e.g. `new_dense`.

fchollet commented on docs/templates/getting-started/faq.md:
```diff
@@ -113,12 +113,39 @@ Note that you will first need to install HDF5 and the Python library h5py, which
+model.add(Dense(2, input_dim=3, name="dense_1"))  # will be loaded
+model.add(Dense(10, name="my_new_embedding"))  # will not be loaded
+
+# load weights from first model
```
specify "will only affect the first layer `dense_1`

fchollet commented on keras/engine/topology.py:
```diff
@@ -2479,14 +2479,27 @@ def save_weights_to_hdf5_group(self, f):
+
+        If `by_name` is False (default) weights are loaded
+        based on the network's topology, meaning the architecture
+        should be the same as when the weights were saved.
```
Note that layers that don't have weights are not taken into account in the topological ordering, so adding or removing layers is fine as long as they don't have weights.

arel commented:
@fchollet this PR is ready.

fchollet commented:
LGTM
